hilti/toolchain/src/compiler/optimizer/passes/flatten-blocks.cc

```diff
@@ -0,0 +1,158 @@
+
+struct Collector : optimizer::visitor::Collector {
+    Collector(Optimizer* optimizer, Variables* variables)
+        : optimizer::visitor::Collector(optimizer), variables(variables) {}
+
+    void operator()(declaration::LocalVariable* local) override { (*variables)[local]; }
+
+    void operator()(expression::Name* name) override {
+        // Since this is a pre-order visitor we will always visit declarations of locals before their use.
+        if ( auto* decl = name->resolvedDeclaration()->tryAs<declaration::LocalVariable>();
+             decl && variables->contains(decl) )
+            (*variables)[decl].insert(name);
+    }
+
+    Variables* variables = nullptr;
+};
+
+Function* containingFunction(Node* n) {
+    Function* function = nullptr;
+    for ( ; n; n = n->parent() ) {
+        if ( auto* f = n->tryAs<Function>() ) {
+            function = f;
+            break;
+        }
+    }
+
+    return function;
+}
+
+struct Mutator : optimizer::visitor::Mutator {
+    Mutator(Optimizer* optimizer) : optimizer::visitor::Mutator(optimizer) {}
+
+    // Tracks names which we will introduce into the parent scope. We don't
+    // rely only on scope lookups since we only reresolve once all locals in a block
+    // have been renamed.
+    std::unordered_map<Function*, std::unordered_map<ID, Declaration*>> used_ids;
+
+    void operator()(statement::Block* block) override {
+        // Only work on blocks which are in other blocks. This excludes e.g.,
+        // blocks which are function bodies, or bodies of `if` or `while`.
+        auto* parent = block->parent()->tryAs<statement::Block>();
+        if ( ! parent )
+            return;
+
+        // We only work on blocks in functions, but not global blocks.
+        hilti::Function* function = containingFunction(block);
+        if ( ! function )
+            return;
+
+        Variables variables;
+        Collector(optimizer(), &variables).run(block);
+
+        // Rename IDs which would clash with existing ones from the parent scope.
+
+        auto& used = used_ids[function];
+
+        for ( auto& [decl, uses] : variables ) {
+            ID id = decl->id();
+            while ( (used.contains(id) && used[id] != decl) || parent->scope()->has(id) )
+                id = ID(id.str() + "_");
+
+            used.emplace(id, decl);
+
+            // No need to trigger ID change if there was no conflict.
+            if ( id == decl->id() )
+                continue;
+
+            recordChange(decl, util::fmt(R"(renaming declaration "%s" -> "%s")", decl->id(), id));
+            decl->setID(id);
+
+            for ( auto* name : uses ) {
+                recordChange(name, util::fmt(R"(renaming reference "%s" -> "%s")", name->id(), id));
+                name->setID(id);
+            }
+        }
+
+        // Fold contents of block into parent.
+        auto contents = parent->children();
```

Comment:
Use `parent->statements()`. Use of `children()` should generally be avoided when working with specific AST node types as that assumes knowledge about the internal node layout.